File: Data Structure/Lab 18/PriorityQueue.c
// WAP to implement Priority Queue.

#include<stdio.h>
#include<stdlib.h>

#define N 7

struct PriorityQueue {
    int data;
    int priority;
};

struct PriorityQueue pq[N];
int PI = 0;

void INSERT(int x,int p){

    if(PI == N){
        printf("Priority Queue Overflow!!");
        return;
    }

    pq[PI].data = x;
    pq[PI].priority = p;
    PI++;
}

void DELETE(){
    if(PI == 0){
        printf("Priority queue underflow!");
        return;
    }

    int highestPI = 0;

    for(int i = 0 ; i < PI ; i++){
        if(pq[i].priority > pq[highestPI].priority){
            highestPI = i;
        }
    }

    int Y = pq[highestPI].data;
    int Ypriority = pq[highestPI].priority;

    printf("Dequeued element : [Value : %d (Priority : %d)]  ",Y,Ypriority);
    printf("\n");

    for(int i = highestPI; i < PI; i++){
        pq[i] = pq[i + 1];
    }
    PI--;
}

void DISPLAY(){

    if(PI == 0){
        printf("Priority Queue Underflow!!");
        return;
    }

    for(int i = 0 ; i < PI; i++){
        printf("[Value : %d(Priority : %d)] ",pq[i].data,pq[i].priority);
    }

    printf("\n");
}

void main() {
    int choice = 0,prio, val, pos,i;

    while (1) {
        printf("\n--- Queue Operation Menu ---\n");
        printf("1-Enqueue an element in priority queue.\n");
        printf("2-Dequeue an element in priority queue.\n");
        printf("3-Display priority queue.\n");

        printf("Enter choice: ");
        scanf("%d", &choice);

        switch (choice) {
            case 1:
                printf("Enter value:");
                scanf("%d",&val);
                printf("Enter priority:");
                scanf("%d",&prio);
                INSERT(val,prio);
                break;
            case 2:
                DELETE();
                break;
            case 3:
                DISPLAY();
                break;
            default:
                printf("Invalid choice.\n");
        }
    }
}

File: Data Structure(Java)/Lab 16/PriorityQueue.java
// WAP to implement Priority Queue.

import java.util.Scanner;

class PriorityQ{
    int data;
    int priority;
}

public class PriorityQueue {
    static int N = 8;

    static PriorityQ PQ[] = new PriorityQ[N];
    static int PI = 0;

    static void INSERT(int x,int p){

        if(PI == N){
            System.out.println("Priority queue overflow!!");
            return;
        }

        PQ[PI] = new PriorityQ();

        PQ[PI].data = x;
        PQ[PI].priority = p;
        PI++;
    }

    static void DELETE(){
        if(PI == 0){
            System.out.println("Priority queue is overflow!");
            return;
        }

        int highestPI = 0;

        for(int i = 0; i < PI ; i++){
            if(PQ[i].priority > PQ[highestPI].priority){
                highestPI = i;
            }
        }

        int Y = PQ[highestPI].data;
        int Ypriority = PQ[highestPI].priority;

        System.out.println("Dequeued element : [Value : " + Y + "(Priority : %d)] " + Ypriority);

        for(int i = highestPI; i < PI; i++){
            PQ[i] = PQ[i + 1];
        }
        PI--;
    }

    static void DISPLAY(){
        if(PI == 0){
            System.out.println("Priority queue underflow!");
            return;
        }

        for(int i = 0; i < PI ; i++){
            System.out.println("[Value : " + PQ[i].data + "(Priority : %d)] " + PQ[i].priority);
        }
    }

    public static void main(String[] args) {
        Scanner sc = new Scanner(System.in);

        while (true) {
            System.out.println("\n--- Priority Queue Operation Menu ---\n");
            System.out.println("1-Enqueue an element.\n");
            System.out.println("2-Dequeue an element.\n");
            System.out.println("3-Display queue.\n");

            System.out.println("Enter choice: ");
            int choice = sc.nextInt();

            switch (choice) {
                case 1:
                    System.out.println("Enter value:");
                    int val = sc.nextInt();
                    System.out.println("Enter priority:");
                    int prio = sc.nextInt();
                    INSERT(val, prio);
                    break;
                case 2:
                    DELETE();
                    break;
                case 3:
                    DISPLAY();
                    break;
                default:
                    System.out.println("Invalid choice.\n");
                    sc.close();
            }
        }
    
    }
}
